src/Nethermind/Nethermind.Consensus/Producers/TxPoolTxSource.cs

```diff
@@ -53,15 +53,65 @@ public IEnumerable<Transaction> GetTransactions(BlockHeader parent, long gasLimi
                 .ThenBy(ByHashTxComparer.Instance); // in order to sort properly and not lose transactions we need to differentiate on their identity which provided comparer might not be doing
 
             IEnumerable<Transaction> transactions = GetOrderedTransactions(pendingTransactions, comparer);
+            IEnumerable<Transaction> blobTransactions = _transactionPool.GetPendingBlobTransactions();
             if (_logger.IsDebug) _logger.Debug($"Collecting pending transactions at block gas limit {gasLimit}.");
 
             int selectedTransactions = 0;
             int i = 0;
-
-            // TODO: removing transactions from TX pool here seems to be a bad practice since they will
-            // not come back if the block is ignored?
             int blobsCounter = 0;
             UInt256 blobGasPrice = UInt256.Zero;
+            List<Transaction>? selectedBlobTxs = null;
+
+            foreach (Transaction blobTx in blobTransactions)
+            {
+                if (BlobGasCalculator.CalculateBlobGas(blobsCounter) == Eip4844Constants.MaxBlobGasPerBlock)
+                {
+                    if (_logger.IsTrace) _logger.Trace($"Declining {blobTx.ToShortString()}, no more blob space. Block already have {blobsCounter} which is max value allowed.");
+                    break;
+                }
+
+                i++;
+
+                bool success = _txFilterPipeline.Execute(blobTx, parent);
+                if (!success) continue;
+
+                if (blobGasPrice.IsZero)
+                {
+                    ulong? excessDataGas = BlobGasCalculator.CalculateExcessBlobGas(parent, _specProvider.GetSpec(parent));
+                    if (excessDataGas is null)
+                    {
+                        if (_logger.IsTrace) _logger.Trace($"Declining {blobTx.ToShortString()}, the specification is not configured to handle shard blob transactions.");
+                        continue;
+                    }
+                    if (!BlobGasCalculator.TryCalculateBlobGasPricePerUnit(excessDataGas.Value, out blobGasPrice))
+                    {
+                        if (_logger.IsTrace) _logger.Trace($"Declining {blobTx.ToShortString()}, failed to calculate data gas price.");
+                        continue;
+                    }
+                }
+
+                int txAmountOfBlobs = blobTx.BlobVersionedHashes?.Length ?? 0;
+
+                if (blobGasPrice > blobTx.MaxFeePerBlobGas)
+                {
+                    if (_logger.IsTrace) _logger.Trace($"Declining {blobTx.ToShortString()}, data gas fee is too low.");
+                    continue;
+                }
+
+                if (BlobGasCalculator.CalculateBlobGas(blobsCounter + txAmountOfBlobs) >
+                    Eip4844Constants.MaxBlobGasPerBlock)
+                {
+                    if (_logger.IsTrace) _logger.Trace($"Declining {blobTx.ToShortString()}, not enough blob space.");
+                    continue;
+                }
```

Comment:
I refactored it a bit to be more readable. Part with `excessDataGas` can't be a part of pipeline, because it is ran only once per block. The same with `TryCalculateBlobGasPricePerUnit`. Actually potential pipeline parts would be `blobGasPrice > blobTx.MaxFeePerBlobGas` and `BlobGasCalculator.CalculateBlobGas(blobsCounter + txAmountOfBlobs) > Eip4844Constants.MaxBlobGasPerBlock` and creating pipeline for 2 filters is probably overkill (or if we will want to add it to standard pipeline, we would need to refactor all old filters to accept `blobGasPrice` and `blobsCOunter` as an argument). IMO current shape with just 2 `if` statements is not pretty, but better than overcomplicating other components